### ISSUE TITLE
Tracking progress for merging the dofhandlers

### ISSUE THREAD

**Issue by kimauth** [Ferrite-FEM/Ferrite.jl#629]
Before merging the dofhandlers, the `MixedDofHandler` must be able to do everything that the `DofHandler` does and should ideally be equally fast. This issue keeps track of how far we've come along that way. 

# Progress tracking
The focus here is to compare how a `MixedDofHandler` performs on a **concrete grid** with all fields on the full domain, i.e. how well `MixedDofHandler` works as a drop-in replacement of `DofHandler`. 

Benchmarking code should go in the comments, ideally one category at a time. However, if a method does not work / does not perform well, open a separate issue about it and reference this one. That way it will be easier to keep an overview and do small reviewable PRs to fix issues.

*Syntax*: Given syntax works and yields correct results for `MixedDofHandler`
*Performance*: performance with `MixedDofHandler` is comparable to `DofHandler` + there is a benchmark of it!

|     | Syntax | Performance |
| --- | --- | --- | 
| | | |
|   **Construction**  |  | ✅  | 
|`DofHandler(grid)` | ✅ |   | 
|`add!(dh, name, dim[, ip])` | ✅  |   | 
|`close!(dh)` | ✅ |  |
| | | |
|   **Basic functionality**  |  |  | 
|`ndofs(dh)` | ✅  | ✅ |
|`ndofs_per_cell(dh[, cell])` | ✅ | ✅ |
|`dof_range(dh, field_name)` | ✅  | ❓ |
|`celldofs(dh, i)` | ✅  | ✅ |
|`celldofs!(dofs, dh, i)` | ✅  | ✅   |
| | | |
|   **Renumbering**  |  |  | 
|`renumber!(dh, order)` | ✅  | ✅  |
|`renumber!(dh, DofOrder.FieldWise())`| ✅  | ✅  |
|`renumber!(dh, DofOrder.ComponentWise())`| ✅   | ✅ |
| | | |
|   **Sparsity pattern**  |  |  | 
|`create_sparsity_pattern(dh)`| ✅  | ✅  |
|`create_sparsity_pattern(dh; coupling)`| ✅  | ✅  |
|`create_symmetric_sparsity_pattern(dh)`| ✅  | ✅ |
|`create_symmetric_sparsity_pattern(dh; coupling)`| ✅ | ✅  |
| | | |
|   **Constraints**  |  |  | 
|`ConstraintHandler(dh)`| ✅ | ✅ |
|    `add!(ch, dbc::Dirichlet)` | ✅ |✅|
|    `add!(ch, ac::AffineConstraint)`| ✅ | ✅  |
|    `add!(ch, pdbc::PeriodicDirichlet)` | ✅ | ✅ |
|    `close!(ch)`| ✅ |✅|
|    `apply_analytical!(a, dh, fieldname, f, cellset)`| ✅  | ❌  |
| | | |
|   **CellIterator**  |  |  | 
|`CellCache(dh)`| ✅ | ✅ |
|    `reinit!(cc, i)`| ✅ | ✅ |
|`CellIterator(dh, cellset)`| ✅  | ❌  |
| | | |
|   **Postprocessing**  |  |  | 
|`get_point_values(ph, dh, dof_values[, fieldname])`| ✅ | ❓ |
|`reshape_to_nodes(dh, u, fieldname)`  | ✅ | ❌ |
|`vtk_point_data(vtk, dh, u)`| ✅ | ❌ |

*Note that all benchmarks might not be equally important. We should discuss based on the results where regressions are acceptable and where they aren't.*

# Benchmarking

A base set-up for benchmarking can look like this:
```julia
using Ferrite

grid = generate_grid(Quadrilateral, (10, 10))

dh = DofHandler(grid)
add!(dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(dh)

mixed_dh = MixedDofHandler(grid)
add!(mixed_dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(mixed_dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(mixed_dh)
```

**Comment by termi-official** [comment]
**Constraints**

```
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (1000, 1000));
∂Ω = union(
    getfaceset(grid, "left"),
    getfaceset(grid, "right"),
    getfaceset(grid, "top"),
    getfaceset(grid, "bottom"),
);
dbc = Dirichlet(:v, ∂Ω, (x, t) -> [0, 0]);
ip_v = Lagrange{2,RefCube,2}();
ip_s = Lagrange{2,RefCube,1}();

function setup_dhclosed(grid, T)
    dh = T(grid)
    add!(dh, :v, 2, Lagrange{2,RefCube,2}()) 
    add!(dh, :s, 1, Lagrange{2,RefCube,1}())
    close!(dh)
    return dh
end

function setup_ch(grid, T)
    dh = setup_dhclosed(grid, T)
    return ConstraintHandler(dh)
end

function setup_ch2(grid, T, dbc)
    dh = setup_dhclosed(grid, T)
    ch = ConstraintHandler(dh)
    add!(ch, dbc)
    return ch
end

@btime ConstraintHandler(dh)  setup=(dh=setup_dhclosed($grid, $DofHandler)) evals=1; # 9.560 μs (12 allocations: 992 bytes)
@btime ConstraintHandler(dh)  setup=(dh=setup_dhclosed($grid, $MixedDofHandler)) evals=1; # 10.130 μs (12 allocations: 992 bytes)

@btime add!(ch, $dbc) setup=(ch = setup_ch($grid, $DofHandler)) evals=1; # 38.962 ms (8108 allocations: 21.91 MiB)
@btime add!(ch, $dbc) setup=(ch = setup_ch($grid, $MixedDofHandler)) evals=1; # 3.617 ms (8124 allocations: 4.04 MiB)

@btime close!(ch) setup=(ch = setup_ch2($grid, $DofHandler, $dbc)) evals=1; # 2.288 s (12071 allocations: 288.21 MiB)
@btime close!(ch) setup=(ch = setup_ch2($grid, $MixedDofHandler, $dbc)) evals=1; # 2.198 s (12071 allocations: 288.21 MiB)
```

**Comment by termi-official** [comment]
**CellIterator**

```julia
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (1000, 1000));
function setup_dhclosed(grid, T)
    dh = T(grid)
    add!(dh, :v, 2, Lagrange{2,RefCube,2}()) 
    add!(dh, :s, 1, Lagrange{2,RefCube,1}())
    close!(dh)
    return dh
end

function setup_cc(grid, T, flags)
    dh = T(grid)
    add!(dh, :v, 2, Lagrange{2,RefCube,2}()) 
    add!(dh, :s, 1, Lagrange{2,RefCube,1}())
    close!(dh)
    return CellCache(dh, flags)
end

@btime CellCache(dh) setup=(dh=setup_dhclosed($grid, $DofHandler)); # 1.062 μs (4 allocations: 480 bytes)
@btime CellCache(dh) setup=(dh=setup_dhclosed($grid, $MixedDofHandler)); # 629.000 ns (4 allocations: 480 bytes)

@btime CellIterator(dh) setup=(dh=setup_dhclosed($grid, $DofHandler)); # 857.000 ns (4 allocations: 480 bytes)
@btime CellIterator(dh) setup=(dh=setup_dhclosed($grid, $MixedDofHandler)); # 3.763 μs (4 allocations: 480 bytes)

@btime reinit!(cc, 1) setup=(cc=setup_cc($grid, $DofHandler, $UpdateFlags(true, true, true))); # 54.031 ns (0 allocations: 0 bytes)
@btime reinit!(cc, 1) setup=(cc=setup_cc($grid, $MixedDofHandler, $UpdateFlags(true, true, true))); # 57.543 ns (0 allocations: 0 bytes)

```

**Comment by termi-official** [comment]
I think bringing down the constructor times for `MixedDofHandler(grid)` down to `DofHandler(grid)` is not worth it. We can move the allocation into close, but it won't change total time.

**Comment by termi-official** [comment]
Not sure if I can fully agree here. I think it makes sense to at least check that we do not have severe performance regression in some simple operations (e.g. due to type instability or unwanted allocs).

**Comment by kimauth** [comment]
**Postprocessing**

```Julia
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (1000, 1000))

dh = DofHandler(grid)
add!(dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(dh)

mixed_dh = MixedDofHandler(grid)
add!(mixed_dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(mixed_dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(mixed_dh)

u = rand(ndofs(dh))

# point evaluation
points = [2*rand(Vec{2})-ones(Vec{2}) for _ in 1:1000]
ph = PointEvalHandler(grid, points)
@btime get_point_values($ph, $dh, $u, $(:v)); # 98.417 μs (28 allocations: 17.44 KiB)
@btime get_point_values($ph, $mixed_dh, $u, $(:v)); # 142.250 μs (31 allocations: 17.56 KiB)

# reshaping from dof-order to nodal order (part of vtk export)
@btime reshape_to_nodes($dh, $u, $(:v)); # 22.863 ms (6 allocations: 22.93 MiB)
@btime reshape_to_nodes($mixed_dh, $u, $(:v)) # 234.516 ms (10 allocations: 22.93 MiB)

# vtk export
filename = joinpath(tempdir(), "test")
@btime vtk_point_data(vtk, $dh, $u) setup=(vtk=vtk_grid($filename, $grid)) evals=1; # 852.222 ms (147 allocations: 52.54 MiB)
@btime vtk_point_data(vtk, $mixed_dh, $u) setup=(vtk=vtk_grid($filename, $grid)) evals=1; # 1.268 s (158 allocations: 52.54 MiB)
```

**Comment by fredrikekre** [comment]
> ```
> @btime celldofs!($dofs, $dh, $15); # 7.675 ns (0 allocations: 0 bytes)
> @btime celldofs!($dofs, $mixed_dh, $15); # 24.072 ns (0 allocations: 0 bytes)
> ```

Those are equally fast for me:
```
@btime celldofs!($dofs, $mixed_dh, $15); # 6.451 ns (0 allocations: 0 bytes)
@btime celldofs!($dofs, $mixed_dh, $15); # 6.393 ns (0 allocations: 0 bytes)

```

**Comment by termi-official** [comment]
Can this be explained by a diff in Julia version or the used machines?

**Comment by kimauth** [comment]
**Sparsity pattern**

```Julia
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (100, 100))

dh = DofHandler(grid)
add!(dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(dh)

mixed_dh = MixedDofHandler(grid)
add!(mixed_dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(mixed_dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(mixed_dh)

# without coupling
@btime create_sparsity_pattern($dh); # 34.338 ms (24 allocations: 246.05 MiB)
@btime create_sparsity_pattern($mixed_dh); # 34.367 ms (24 allocations: 246.05 MiB)

@btime create_symmetric_sparsity_pattern($dh); # 18.820 ms (24 allocations: 130.69 MiB)
@btime create_symmetric_sparsity_pattern($mixed_dh); # 18.739 ms (24 allocations: 130.69 MiB)

# with coupling
field_coupling = [true false; false true]

@btime create_sparsity_pattern($dh; coupling=$field_coupling); # 27.862 ms (26 allocations: 175.80 MiB)
# create_sparsity_pattern(mixed_dh; coupling=field_coupling); # TODO: missing

@btime create_symmetric_sparsity_pattern($dh; coupling=$field_coupling); # 16.695 ms (26 allocations: 119.83 MiB)
# create_sparsity_pattern(mixed_dh; coupling=field_coupling); # TODO: missing
```

**Comment by kimauth** [comment]
**CellIterator II** 

> ```
> @btime CellCache(dh) setup=(dh=setup_dhclosed($grid, $DofHandler)); # 1.062 μs (4 allocations: 480 bytes)
> @btime CellCache(dh) setup=(dh=setup_dhclosed($grid, $MixedDofHandler)); # 629.000 ns (4 allocations: 480 bytes)
> ```

Can't reproduce the difference in constructing `CellCache` 
```Julia
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (1000, 1000));

dh = DofHandler(grid)
add!(dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(dh)

mixed_dh = MixedDofHandler(grid)
add!(mixed_dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(mixed_dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(mixed_dh)

@btime CellCache($dh); # 74.359 ns (4 allocations: 480 bytes)
@btime CellCache($mixed_dh); # 73.665 ns (4 allocations: 480 bytes)

@btime CellIterator($dh); # 74.820 ns (4 allocations: 480 bytes)
@btime CellIterator($mixed_dh); # 9.708 μs (4 allocations: 480 bytes)

cc_dh = CellCache(dh);
cc_mixed_dh = CellCache(mixed_dh)
@btime reinit!($cc_dh, $1); # 20.938 ns (0 allocations: 0 bytes)
@btime reinit!($cc_mixed_dh, $1); # 22.503 ns (0 allocations: 0 bytes)
```

**Comment by kimauth** [comment]
**Constraints: apply_analytical, affine constraints, periodic bcs**

```julia
f(x) = x ⋅ x
u = zeros(ndofs(dh))

@btime apply_analytical!($u, $dh, $(:s), $f); # 25.878 ms (14 allocations: 2.27 KiB)
@btime apply_analytical!($u, $mixed_dh, $(:s), $f); # 323.590 ms (76 allocations: 34.50 MiB)

lc = AffineConstraint(1, [2 => 5.0, 3 => 3.0], 1.0)
@btime add!(ch, $lc) setup=(ch = ConstraintHandler($dh)); # 11.596 ns (0 allocations: 0 bytes)
@btime add!(ch, $lc) setup=(ch = ConstraintHandler($mixed_dh)); # 11.177 ns (0 allocations: 0 bytes)

φ(x) = x - Vec{2}((1.0, 0.0))
face_mapping = collect_periodic_faces(grid, "left", "right", φ)
pdbc = PeriodicDirichlet(:v, face_mapping, [1, 2])
# Add the constraint to the constraint handler
@btime add!(ch, $pdbc) setup=(ch=ConstraintHandler($dh)); # 737.750 μs (4091 allocations: 1.24 MiB)
@btime add!(ch, $pdbc) setup=(ch=ConstraintHandler($mixed_dh)); # 735.458 μs (4091 allocations: 1.24 MiB)
```

The difference in `apply_analytical` seems to a significant part be caused by poor memory alignment and thus should mostly be fixed by fixing #631 .

**Comment by kimauth** [comment]
**Renumbering**

```Julia
using Ferrite
using BenchmarkTools

grid = generate_grid(Quadrilateral, (100, 100))

dh = DofHandler(grid)
add!(dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(dh)

mixed_dh = MixedDofHandler(grid)
add!(mixed_dh, :v, 2, Lagrange{2,RefCube,2}()) # quadratic vector field
add!(mixed_dh, :s, 1, Lagrange{2,RefCube,1}()) # linear scalar field
close!(mixed_dh)

@btime renumber!($dh, $(ndofs(dh):-1:1)) # 596.500 μs (4 allocations: 22.56 KiB)
@btime renumber!($mixed_dh, $(ndofs(mixed_dh):-1:1)) # 598.459 μs (4 allocations: 22.56 KiB)

@btime renumber!($dh, $(DofOrder.FieldWise())) # 6.075 ms (91 allocations: 5.87 MiB)
# renumber!(mixed_dh, DofOrder.FieldWise()) # TODO: missing

@btime renumber!($dh, $(DofOrder.ComponentWise())) # 5.902 ms (111 allocations: 4.33 MiB)
# renumber!(mixed_dh, DofOrder.ComponentWise()) # TODO: missing
```

**Comment by fredrikekre** [comment]
After #660 I think this issue can be closed since the MixedDofHandler is now equally as performant (at least where it really matters). :tada:
